### PR TITLE
PML implementation overhaul

### PR DESCRIPTION
**Pull request by galcerte**
This PR tackles issue #172. The initial commit on here only fixes the first two of the three points mentioned there. The last point on that issue is what will probably take the most time to complete, that is, updating the fields once across all PMLs.

Even though now the PML object updates the fields according to literature, in my testing, the PML is still behaving rather weirdly, even if no PML volumes in different axis overlap. This is the script I am using to test this out: [https://gist.github.com/galcerte/d062bc8446a766be5e03a2d8f4f5fecc](https://gist.github.com/galcerte/d062bc8446a766be5e03a2d8f4f5fecc). Maybe it's worth including this script in the `checks/` folder?

Other things that still need to be done is to check if the PMLs work fine with lossy materials inside them.

### CODE REVIEW

**Comment by ymahlau** [comment]
I will take a look at this later. I think for implementing non-overlapping PML it would be best to generalize the PML implementation to multiple directions (x, y, z, either positive / negative), which can be specified by the user. This would allow specifying a separate PML object for the corner and sides. Then the get_boundary_from_config can automatically generate the appropriate objects (same pml as before on the side, separate pml object for corners). Specifying the position of these objects by hand would probably be tedious, but should be easy to automate.

**Comment by ymahlau** [comment]
As discussed in #84, the updates in the pml are not carried out in parallel, because the pmls return the complete array, which makes it difficult for the jit optimizer to identify that the updates can actually be performed in parallel. There are ways to do this better (constructing a new array from the updated pmls and the non-updated regions through concatenation), but this may be a lot of work to implement.  Could you add the new implementation and also keep the old one such that a user can decide if they want the higher accuracy or faster runtime?

**Comment by galcerte** [comment]
I was thinking of another way to go about it; instead of updating 26 times, what about taking all the grid slices, taking in all the alpha, kappa, and sigmas (as well as the bE and bH, cB and cH coefficients; basically everything in PMLBoundaryState), concatenating _those_ and updating _just once_? The problem with this is that the updating would probably have to be done elsewhere, as one single `update_E` or `update_H` method in each PML object does not have the full picture by accessing their own object's attributes; probably a different helper function distinct from all those methods would do the trick. Then each PML object would be in charge of just storing the previous state in PMLBoundaryState, passing them to the function, and letting it concatenate all those arrays and update the fields using that.

Or maybe it would be worth making just 1 PML object for all PML boundaries. That way the `update_E` and `update_H` interface would be intact, and its usage in `update.py`. But that would need more work in `initialization.py`.

> Could you add the new implementation and also keep the old one such that a user can decide if they want the higher accuracy or faster runtime?

I can do that, but honestly I also think it's worth putting in the effort to make this implementation be both accurate and fast.

**Comment by ymahlau** [comment]
Yes, if you can manage to make the implementation both fast and accurate then we do not need the old parts anymore. However, I think it would be difficult to make this work in all cases, especially when using multiple GPU. For these cases we should keep the old implementation as an option.

The options you are suggesting all eventually come down to the same thing I was saying above:
- you will need to give the pml objects the individual slices of the E/H field and retrieve the updated E/H slices. jax will do that in parallel automatically.
- create a 3d array of the original simulation volume size from these arrays using jnp.concatenate and use the old E/H field where no updates happened (this should be one big slice in the middle of the simulation volume. If there is a combination of periodic and pml boundaries this will need to be adjusted.
